### DISCUSSION TITLE
More flexible routing for react

### DISCUSSION THREAD

**Discussion by laran** [activej/activej#164] [Q&A]
I followed the instructions on how to use the StaticServlet to serve a react app. It works fine when I enter the app from "/". But when I go directly to "/contact" or any other specific URL the routing seems to fail with a 404.

Is there a more flexible way to provide routing for a react app, so that a user could go directly to any URL and have react serve it up properly?

**Comment by eduard-vasinskyi** [comment]
Hi, @laran

When you go directly to “/contact” path you send a request to [StaticServlet](https://github.com/activej/activej/blob/master/core-http/src/main/java/io/activej/http/StaticServlet.java) to load a virtual path “/contact” which only exists as a route in your react application. You need to delegate the processing of such routes directly to react app. To do this you can map all not found resources to “index.html”. This way a react app will do the actual routing to “/contact” or any other react route that you have defined.

Modify `@Provides` method for `StaticServlet` by adding a `StaticServlet#withMappingNotFoundTo` method call and pass “index.html” as a fallback resource.

```java
@Provides
AsyncServlet servlet(Executor executor) {
  return StaticServlet.ofClassPath(executor, "build")
        .withIndexHtml()
        .withMappingNotFoundTo("index.html");
}
```
